File: src/db/store.js
/**
 * PostgreSQL-based data store.
 * Implements the same interface as the JSON store for seamless transition.
 */

const { query } = require('./db');

const TABLES = {
  USERS:                'users',
  TEACHERS:             'teachers',
  STUDENTS:             'students',
  PRINCIPALS:           'principals',
  INSTITUTIONS:         'institutions',
  USER_ROLES:           'user_roles',
  DEPARTMENTS:          'departments',
  SUBJECTS:             'subjects',
  SHIFTS:               'shifts',
  FACE_TEMPLATES:       'face_embedding',
  ENROLLMENT_SESSIONS:  'face_enrollment',
  FACE_ENROLLMENT:      'face_enrollment',
  ENROLLMENT_SAMPLES:   'face_enrollment_images',
  ATTENDANCE_LOGS:      'teacher_attendance',
  ATTENDANCE_SUMMARY:   'attendance_daily_summary',
  GEOFENCES:            'geofences',
  ATTENDANCE_POLICIES:  'attendance_policies',
  ATTENDANCE_EXCEPTIONS:'attendance_exceptions',
  DEVICE_REGISTRY:      'user_devices',
  SYNC_QUEUE:           'sync_queue_log',
  AUDIT_LOGS:           'audit_logs',
  TEACHER_SUBJECT_MAPPING: 'teacher_subject_mapping',
  ATTENDANCE_SESSIONS:     'attendance_sessions',
  STUDENT_ATTENDANCE:      'student_attendance',
  TEACHER_ATTENDANCE:      'teacher_attendance',
  FACE_ENROLLMENT:         'face_enrollment',
  FACE_EMBEDDING:          'face_embedding',
  FACE_ENROLLMENT_IMAGES:  'face_enrollment_images',
  FACE_RECOGNITION_LOGS:   'face_recognition_logs',
  AI_SESSION_PROCESSING:   'ai_session_processing',
  LEAVE_TYPES:             'leave_types',
  LEAVE_APPLICATIONS:      'leave_applications',
  ACADEMIC_CALENDAR:       'academic_calendar',
  USER_DEVICES:            'user_devices',
  SYSTEM_SETTINGS:         'system_settings',
  NOTIFICATIONS:           'notifications',
  COURSES:                 'courses',
  SEMESTERS:               'semesters',
  CLASSES:                 'classes',

};
// ── Generic helpers (Mapped to SQL) ──────────────────────────────────────────

async function getAll(table) {
  const { rows } = await query(`SELECT * FROM ${table}`);
  return rows;
}

async function getById(table, idOrField, id) {
  const field = id === undefined ? 'id' : idOrField;
  const val   = id === undefined ? idOrField : id;
  const { rows } = await query(`SELECT * FROM ${table} WHERE ${field} = $1`, [val]);
  return rows[0] || null;
}

/**
 * findOne / findMany are tricky with SQL using predicates.
 * We'll implement a simplified version that handles common cases (id, employee_id, etc.)
 * For complex queries, controllers should ideally use a specialized query function.
 */
async function findOne(table, predicateOrQuery) {
  if (typeof predicateOrQuery === 'function') {
    // FALLBACK: If a function is passed, we fetch ALL and filter (In-memory, suboptimal)
    // This allows existing code to work while migrating.
    const all = await getAll(table);
    return all.find(predicateOrQuery) || null;
  }
  
  // If it's an object, we build a simple WHERE clause
  const keys = Object.keys(predicateOrQuery);
  const where = keys.map((k, i) => `${k} = $${i + 1}`).join(' AND ');
  const { rows } = await query(`SELECT * FROM ${table} WHERE ${where} LIMIT 1`, Object.values(predicateOrQuery));
  return rows[0] || null;
}

async function findMany(table, predicateOrQuery) {
  if (typeof predicateOrQuery === 'function') {
    const all = await getAll(table);
    return all.filter(predicateOrQuery);
  }

  const keys = Object.keys(predicateOrQuery);
  if (keys.length === 0) return getAll(table);

  const where = keys.map((k, i) => `${k} = $${i + 1}`).join(' AND ');
  const { rows } = await query(`SELECT * FROM ${table} WHERE ${where}`, Object.values(predicateOrQuery));
  return rows;
}

async function insert(table, record) {
  const keys   = Object.keys(record);
  const values = Object.values(record);
  const cols   = keys.join(', ');
  const placeholders = keys.map((_, i) => `$${i + 1}`).join(', ');

  const sql = `INSERT INTO ${table} (${cols}) VALUES (${placeholders}) RETURNING *`;
  const { rows } = await query(sql, values);
  return rows[0];
}

async function update(table, idOrField, idOrChanges, changes) {
  let field, id, finalChanges;
  
  if (changes === undefined) {
    field = 'id';
    id = idOrField;
    finalChanges = idOrChanges;
  } else {
    field = idOrField;
    id = idOrChanges;
    finalChanges = changes;
  }

  const keys = Object.keys(finalChanges);
  if (keys.length === 0) return getById(table, field, id);

  const setClause = keys.map((k, i) => `${k} = $${i + 2}`).join(', ');
  const values = [id, ...Object.values(finalChanges)];

  const sql = `UPDATE ${table} SET ${setClause} WHERE ${field} = $1 RETURNING *`;
  const { rows } = await query(sql, values);
  return rows[0] || null;
}

async function remove(table, idOrField, id) {
  const field = id === undefined ? 'id' : idOrField;
  const val   = id === undefined ? idOrField : id;
  const { rowCount } = await query(`DELETE FROM ${table} WHERE ${field} = $1`, [val]);
  return rowCount > 0;
}

module.exports = {
  TABLES,
  getAll, getById, findOne, findMany,
  insert, update, remove,
};

File: face_attendance_dart/bin/database/store.dart
import 'package:postgres/postgres.dart';
import 'database_connection.dart';

class Store {
  /// Fetches all records from a table.
  static Future<List<Map<String, dynamic>>> getAll(String table) async {
    final pool = await DatabaseConnection.pool;
    final result = await pool.execute('SELECT * FROM $table');
    return result.map((row) => row.toColumnMap()).toList();
  }

  /// Fetches a single record by ID or a specific field.
  static Future<Map<String, dynamic>?> getById(String table, dynamic id, {String field = 'id'}) async {
    final pool = await DatabaseConnection.pool;
    final result = await pool.execute(
      Sql.named('SELECT * FROM $table WHERE $field = @id LIMIT 1'),
      parameters: {'id': id},
    );
    return result.isEmpty ? null : result.first.toColumnMap();
  }

  /// Finds a single record matching the predicate.
  static Future<Map<String, dynamic>?> findOne(String table, Map<String, dynamic> predicate) async {
    final pool = await DatabaseConnection.pool;
    if (predicate.isEmpty) return null;

    final keys = predicate.keys.toList();
    final where = keys.map((k) => '$k = @$k').join(' AND ');

    final result = await pool.execute(
      Sql.named('SELECT * FROM $table WHERE $where LIMIT 1'),
      parameters: predicate,
    );
    return result.isEmpty ? null : result.first.toColumnMap();
  }

  /// Finds multiple records matching the predicate.
  static Future<List<Map<String, dynamic>>> findMany(String table, Map<String, dynamic> predicate) async {
    final pool = await DatabaseConnection.pool;
    if (predicate.isEmpty) return getAll(table);

    final keys = predicate.keys.toList();
    final where = keys.map((k) => '$k = @$k').join(' AND ');

    final result = await pool.execute(
      Sql.named('SELECT * FROM $table WHERE $where'),
      parameters: predicate,
    );
    return result.map((row) => row.toColumnMap()).toList();
  }

  /// Inserts a new record into the table.
  static Future<Map<String, dynamic>> insert(String table, Map<String, dynamic> record) async {
    final pool = await DatabaseConnection.pool;
    final keys = record.keys.toList();
    final cols = keys.join(', ');
    final placeholders = keys.map((k) => '@$k').join(', ');

    final result = await pool.execute(
      Sql.named('INSERT INTO $table ($cols) VALUES ($placeholders) RETURNING *'),
      parameters: record,
    );
    return result.first.toColumnMap();
  }

  /// Updates an existing record.
  static Future<Map<String, dynamic>?> update(String table, dynamic id, Map<String, dynamic> changes, {String field = 'id'}) async {
    final pool = await DatabaseConnection.pool;
    if (changes.isEmpty) return getById(table, id, field: field);

    final keys = changes.keys.toList();
    final setClause = keys.map((k) => '$k = @$k').join(', ');
    
    final parameters = Map<String, dynamic>.from(changes);
    parameters['id_val_where'] = id;

    // Use a different placeholder for the WHERE clause to avoid collision
    final result = await pool.execute(
      Sql.named('UPDATE $table SET $setClause WHERE $field = @id_val_where RETURNING *'),
      parameters: parameters,
    );
    return result.isEmpty ? null : result.first.toColumnMap();
  }

  /// Removes a record from the table.
  static Future<bool> remove(String table, dynamic id, {String field = 'id'}) async {
    final pool = await DatabaseConnection.pool;
    final result = await pool.execute(
      Sql.named('DELETE FROM $table WHERE $field = @id'),
      parameters: {'id': id},
    );
    return result.affectedRows > 0;
  }
}
